pkg/blockbuilder/blockbuilder.go

```diff
@@ -0,0 +1,513 @@
+
+package blockbuilder
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"flag"
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/dskit/backoff"
+	"github.com/grafana/dskit/services"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/thanos-io/objstore"
+	"github.com/twmb/franz-go/pkg/kadm"
+	"github.com/twmb/franz-go/pkg/kerr"
+	"github.com/twmb/franz-go/pkg/kgo"
+	"github.com/twmb/franz-go/plugin/kprom"
+
+	"github.com/grafana/mimir/pkg/storage/bucket"
+	"github.com/grafana/mimir/pkg/storage/ingest"
+	storagetsdb "github.com/grafana/mimir/pkg/storage/tsdb"
+	"github.com/grafana/mimir/pkg/util/validation"
+)
+
+type BlockBuilder struct {
+	services.Service
+
+	cfg         Config
+	logger      log.Logger
+	register    prometheus.Registerer
+	limits      *validation.Overrides
+	kafkaClient *kgo.Client
+	bucket      objstore.Bucket
+
+	parts []int32
+	// fallbackOffset is the low watermark from where a partition which doesn't have a commit will be consumed.
+	// It can be either a timestamp or the kafkaStartOffset (the latter isn't supported still).
+	// TODO(v): to support setting it to kafkaStartOffset we need to rework how a lagging cycle is chopped into time frames.
+	fallbackOffset int64
+
+	metrics blockBuilderMetrics
+}
+
+func New(
+	cfg Config,
+	logger log.Logger,
+	reg prometheus.Registerer,
+	limits *validation.Overrides,
+) (*BlockBuilder, error) {
+	b := &BlockBuilder{
+		cfg:      cfg,
+		logger:   logger,
+		register: reg,
+		limits:   limits,
+		metrics:  newBlockBuilderMetrics(reg),
+	}
+
+	instanceID, err := parseIDFromInstance(b.cfg.InstanceID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse instance id: %w", err)
+	}
+
+	var ok bool
+	b.parts, ok = b.cfg.PartitionAssignment[instanceID]
+	if !ok {
+		return nil, fmt.Errorf("instance id %d must be present in partition assignment", instanceID)
+	}
+
+	bucketClient, err := bucket.NewClient(context.Background(), cfg.BlocksStorageConfig.Bucket, "block-builder", logger, reg)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create the bucket client: %w", err)
+	}
+	b.bucket = bucketClient
+
+	b.Service = services.NewBasicService(b.starting, b.running, b.stopping)
+
+	return b, nil
+}
+
+func parseIDFromInstance(instanceID string) (int, error) {
+	splits := strings.Split(instanceID, "-")
+	if len(splits) == 0 {
+		return 0, fmt.Errorf("malformed instance id %s", instanceID)
+	}
+
+	partID, err := strconv.Atoi(splits[len(splits)-1])
+	if err != nil {
+		return 0, fmt.Errorf("parse instance id %s, expect sequence number: %w", instanceID, err)
+	}
+
+	return partID, nil
+}
+
+func (b *BlockBuilder) starting(ctx context.Context) (err error) {
+	defer func() {
+		if err != nil {
+			if b.kafkaClient != nil {
+				b.kafkaClient.Close()
+			}
+		}
+	}()
+
+	// Empty any previous artifacts.
+	if err := os.RemoveAll(b.cfg.BlocksStorageConfig.TSDB.Dir); err != nil {
```

Comment:
I think we should use dedicated folder (specified in the block builder Config) instead of the TSDB one. TSDB one is the one used by ingesters and I don't want to run into any conflict when running, for example, the monolithic mode during development / local testing.